# Task Exercise 4.9 Book 1. Instacart Customers Data Cleaning

## The senior Instacart officers have given you a new data set of customer information to go along with your product and order data. In part 2 of the task, you’ll need to incorporate this new data set into your project. In part 2, you’ll create some visualizations, conduct some exploratory analysis, and begin wrapping up everything you’ve done in this Achievement in preparation for the final task in the next Exercise, where you’ll write up a report for your client.

### Due to the size of the dataframes involved, this exercise was broken into multiple notebooks to conserve memory.  Book 1. focused on importing the 'customers.csv' dataframe, verifying, cleaning and modifying it, and exporting it as 'customers_checked.csv'.


## This script contains the following points:

## 1. Import and verify 'customers.csv' dataframe

## 2. Renaming columns

### 'FIrst Name' to 'first_name
### 'Surnam' to 'surname'
### 'Gender to 'gender'
### 'STATE' to 'state'
### 'Age' to 'age'
### 'n_dependants' to 'num_of_dependants'
### 'fam_status' to "family_status'

## 3. Finding duplicates

## 4. Convert 'user_id' to object to facilitate merger with 'orders_products_merged' dataframe

## 5. Export  'customers_checked.csv'


In [1]:
# import libraries

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
path = r'C:\Users\howl6\OneDrive\Certificates\CareerFoundry\Coursework\Data_Immersion\Chapter 4\Instacart Basket Analysis'

In [3]:
df_cust = pd.read_csv(os.path.join(path,'02_Data','Original_Data', '4.9_customers', 'customers.csv'))

In [4]:
# view top 20 rows

df_cust.head(20)

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0     26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1     33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2     65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3    125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4    130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5    133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6    152052      Chris    Walton    Male         Montana   20    1/1/2017   
7    168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8     69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9     82820      Shawn     Chung    Male        Virginia   26    1/1/2017   
10    26441     Gloria  Stafford  Female          Nevada   19    1/1/2017   
11    33835      Roger     Casey    Male        Nebraska   45    1/1/2017   
12    75032    Cynthia  Mcdaniel  Female         Georgia   75    1/1/2017   
13       51  Stephanie    Jensen  Female         Wyoming   37    1/1/2017   
14      516      Peter      Hunt    Male        Colorado   51    1/1/2017   
15     6965    Brandon      Mata    Male    North Dakota   70    1/1/2017   
16     8794        Amy      Mack  Female        Nebraska   30    1/1/2017   
17     9944      Brian    Oneill    Male       Wisconsin   26    1/1/2017   
18    13313     Victor      Moss    Male          Alaska   31    1/1/2017   
19    17335    Patrick    Fowler    Male         Vermont   65    1/1/2017   

    n_dependants                        fam_status  income  
0              3                           married  165665  
1              0                            single   59285  
2              2                           married   99568  
3              0                            single   42049  
4              1                           married   40374  
5              2                           married   49643  
6              0                            single   61746  
7              0                            single   63712  
8              1                           married  162432  
9              2                           married   32072  
10             2  living with parents and siblings   43443  
11             2                           married   47988  
12             1                           married   57835  
13             1                           married   71190  
14             2                           married  146559  
15             0                  divorced/widowed  100091  
16             3                           married   57096  
17             1                           married   91826  
18             2                           married   91111  
19             1                           married   34514

In [5]:
# view bottom 35 rows

df_cust.tail(35)

user_id   First Name      Surnam  Gender           STATE  Age  \
206174   134553        Ralph      Avalos    Male         Indiana   25   
206175   167749      Deborah     Farrell  Female         Florida   28   
206176   186595         Ruth  Cunningham  Female     Mississippi   38   
206177   199732    Katherine      Abbott  Female            Iowa   71   
206178   138442       Gloria    Cantrell  Female  North Carolina   26   
206179   177599        Donna        Ruiz  Female          Kansas   71   
206180    48091       Jeremy      Willis    Male   West Virginia   37   
206181   192077        Susan        Nash  Female         Georgia   54   
206182   180181         Carl     Bridges    Male   West Virginia   77   
206183    79560       Steven      Sutton    Male         Wyoming   78   
206184    61944       Walter        Park    Male            Ohio   54   
206185   149397         Tina          Wu  Female        Kentucky   22   
206186   107955       Victor      Guzman    Male    Pennsylvania   41   
206187   163443          Roy      Harvey    Male    Pennsylvania   37   
206188    93669     Patricia        Cole  Female        Maryland   74   
206189   193828      Russell      Travis    Male  North Carolina   46   
206190   197067        Kathy        Bell  Female         Arizona   42   
206191     7177      Russell   Zimmerman    Male     Mississippi   71   
206192    61888       Joshua      Guerra    Male      New Jersey   37   
206193   103412       Willie     Goodman    Male        Michigan   46   
206194   189337        Shawn        Wood    Male      New Jersey   56   
206195   205766       Amanda       Hodge  Female          Oregon   18   
206196   139950       Gloria      Murray  Female        Colorado   45   
206197    74598  Christopher   Velazquez    Male       Minnesota   52   
206198    83573       Gloria      Murray  Female        Michigan   28   
206199   179673         Adam  Villanueva    Male         Wyoming   77   
206200   192637        Kelly    Petersen  Female         Florida   47   
206201    75529        Kathy       Avila  Female   West Virginia   25   
206202   157533          Ned       Stark    Male            Utah   53   
206203   122741         Carl       Weber    Male        Michigan   24   
206204   168073         Lisa        Case  Female  North Carolina   44   
206205    49635       Jeremy     Robbins    Male          Hawaii   62   
206206   135902        Doris    Richmond  Female        Missouri   66   
206207    81095         Rose     Rollins  Female      California   27   
206208    80148      Cynthia       Noble  Female        New York   55   

       date_joined  n_dependants                        fam_status  income  
206174    4/1/2020             1                           married   64482  
206175    4/1/2020             1                           married   30169  
206176    4/1/2020             1                           married   92727  
206177    4/1/2020             1                           married   31019  
206178    4/1/2020             1                           married   46199  
206179    4/1/2020             3                           married  119306  
206180    4/1/2020             3                           married   31048  
206181    4/1/2020             3                           married  160664  
206182    4/1/2020             2                           married  167232  
206183    4/1/2020             3                           married  101764  
206184    4/1/2020             0                            single   58050  
206185    4/1/2020             1                           married   32242  
206186    4/1/2020             3                           married  136624  
206187    4/1/2020             0                            single   82652  
206188    4/1/2020             1                           married  122136  
206189    4/1/2020             1                           married  160483  
206190    4/1/2020             0                            single  11

In [6]:
# view number of rows, columns

df_cust.shape

(206209, 10)

In [7]:
# list of columns

df_cust.columns

Index(['user_id', 'First Name', 'Surnam', 'Gender', 'STATE', 'Age',
       'date_joined', 'n_dependants', 'fam_status', 'income'],
      dtype='object')

In [8]:
# view descriptive statistics

df_cust.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

### A review of the descriptive statistics does not suggest any significant outliers.

In [9]:
# view datatypes

df_cust.dtypes

user_id          int64
First Name      object
Surnam          object
Gender          object
STATE           object
Age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object

### Variable types all seem appropriate.

In [10]:
# looking for missing values (NaN)

df_cust['user_id'].value_counts(dropna = False)

2049      1
167163    1
187633    1
181490    1
183539    1
         ..
150044    1
147997    1
154142    1
152095    1
2047      1
Name: user_id, Length: 206209, dtype: int64

In [11]:
df_cust['First Name'].value_counts(dropna = False)

NaN        11259
Marilyn     2213
Barbara     2154
Todd        2113
Jeremy      2104
           ...  
Merry        197
Eugene       197
Garry        191
David        186
Ned          186
Name: First Name, Length: 208, dtype: int64

In [12]:
df_cust['date_joined'].value_counts(dropna = False)

9/17/2018     213
2/10/2018     212
4/1/2019      211
9/21/2019     211
12/19/2017    210
             ... 
9/1/2018      141
1/22/2018     140
11/24/2017    139
7/18/2019     138
8/6/2018      128
Name: date_joined, Length: 1187, dtype: int64

In [13]:
df_cust['income'].value_counts(dropna = False)

95891     10
57192     10
95710     10
94809      9
97532      9
          ..
139481     1
152861     1
464181     1
228664     1
28658      1
Name: income, Length: 108012, dtype: int64

In [14]:
df_cust['Gender'].value_counts(dropna = False)

Male      104067
Female    102142
Name: Gender, dtype: int64

In [15]:
df_cust.isnull().sum()

user_id             0
First Name      11259
Surnam              0
Gender              0
STATE               0
Age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64

In [16]:
df_nan = df_cust[df_cust['First Name'].isnull() == True]

In [17]:
df_nan

user_id First Name    Surnam  Gender         STATE  Age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        n_dependants fam_status  income  
53                 2    married   41709  
73                 0     single   82518  
82                 3    married  117099  
99                 1    married  155673  
105                3    married  151819  
...              ...        ...     ...  
206038             3    married   87783  
206044             2    married  108488  
206090             1    married   45275  
206162             0     single   56325  
206171             2    married   59222  

[11259 rows x 10 columns]

### A review of column values provides variable counts and identifies 11259 missing first names.


## 2. Renaming columns 

### 'FIrst Name' to 'first_name
### 'Surnam' to 'surname'
### 'Gender to 'gender'
### 'STATE' to 'state'
### 'Age' to 'age'
### 'n_dependants' to 'num_of_dependants'
### 'fam_status' to "family_status'

In [18]:
df_cust.rename(columns = {'First Name' : 'first_name'}, inplace = True)

In [19]:
df_cust.rename(columns = {'Surnam' : 'surname'}, inplace = True)

In [20]:
df_cust.rename(columns = {'Gender' : 'gender'}, inplace = True)

In [21]:
df_cust.rename(columns = {'STATE' : 'state'}, inplace = True)

In [22]:
df_cust.rename(columns = {'Age' : 'age'}, inplace = True)

In [23]:
df_cust.rename(columns = {'n_dependants' : 'num_of_dependants'}, inplace = True)

In [24]:
df_cust.rename(columns = {'fam_status' : 'family_status'}, inplace = True)

In [25]:
df_cust.head()

user_id first_name   surname  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   num_of_dependants family_status  income  
0                  3       married  165665  
1                  0        single   59285  
2                  2       married   99568  
3                  0        single   42049  
4                  1       married   40374

## 3. Finding duplicates

In [26]:
df_dups = df_cust[df_cust.duplicated()]

In [27]:
df_dups

Empty DataFrame
Columns: [user_id, first_name, surname, gender, state, age, date_joined, num_of_dependants, family_status, income]
Index: []

## There are no duplicate variables in the dataframe.

## 4. Convert 'user_id' to object to facilitate merger with 'orders_products_merged' dataframe

In [28]:
df_cust['user_id'] = df_cust['user_id'].astype('str')

In [29]:
df_cust.dtypes

user_id              object
first_name           object
surname              object
gender               object
state                object
age                   int64
date_joined          object
num_of_dependants     int64
family_status        object
income                int64
dtype: object

## 5. Export  'customers_checked.csv'

In [30]:
df_cust.to_csv(os.path.join(path, '02_Data','Prepared_Data', 'customers_checked.csv'))